### PR TITLE
Add option to disable jquery-migrate + WP 5.5 updates

### PR DESCRIPTION
**Pull request by cssjoe**

### CODE REVIEW

File: ConfigKeys.php [hunk 1]
```diff
@@ -2417,7 +2417,10 @@
 		'default' => '',
 		'master_only' => true
 	),
-
+	'jquerymigrate.disabled' => array(
```
Comment (bwmarkle) [inline ConfigKeys.php:2420]: why

File: Generic_Plugin.php [hunk 2]
```diff
@@ -84,6 +84,31 @@ function run() {
 					'ob_callback'
 				) );
 		}
+
+		// If WP < 5.5, disable jquery-migrate on the front-end, if configured.
```
Comment (bwmarkle) [inline Generic_Plugin.php:88]: why

File: Generic_Plugin.php [hunk 3]
```diff
@@ -84,6 +84,31 @@ function run() {
...
+		if ( ! is_admin() && version_compare( get_bloginfo( 'version' ), '5.5', '<' ) && $this->_config->get_boolean( 'jquerymigrate.disabled' ) ) {
+			add_action( 'wp_default_scripts', array( $this, 'disable_jquery_migrate' ) );
+		}
+	}
+
+	/**
+	 * Disable jquery-migrate.
```
Comment (bwmarkle) [inline Generic_Plugin.php:95]: why

File: inc/options/general.php [hunk 4]
```diff
@@ -481,6 +481,17 @@ class="w3tc-ignore-change" type="text"
 				</th>
 			</tr>
 			<?php
+			if ( version_compare( get_bloginfo( 'version' ), '5.5', '<' ) ) {
```
Comment (bwmarkle) [inline inc/options/general.php:484]: So we're adding a new setting for users to disable jquery migrate on the front end? I wasn't really aware what was in this ticket.

File: Generic_Plugin.php (hunk 2 above)
Comment (cssjoe) [reply Generic_Plugin.php]: I removed the version check.

File: Generic_Plugin.php (hunk 3 above)
Comment (cssjoe) [reply Generic_Plugin.php:95]: It was a feature request that was marked Next.

File: inc/options/general.php (hunk 4 above)
Comment (cssjoe) [reply inc/options/general.php]: Yes, see the link in this PR.  I removed the version check.  https://make.wordpress.org/core/2020/06/29/updating-jquery-version-shipped-with-wordpress/

**Comment by maxicus** [comment]
The change is fine by idea, just the logic itself shouldn't be in general module, but user experience one.
In general nothing should be in general module :)

File: Extension_CloudFlare_Page_View.js [hunk 5]
```diff
@@ -33,7 +33,7 @@ jQuery(function($) {
 
 
 		.on('click', '.w3tc_cloudflare_zone_page', function() {
-			var page = jQuery(this).attr('data-page');
+			var page = jQuery(this).prop('data-page');
```
Comment (maxicus) [inline Extension_CloudFlare_Page_View.js:36]: That and all similar looks like a bug.
prop shouldnt return value for non-DOM properties.

Did you check it works?

File: Generic_Plugin.php [hunk 6]
```diff
@@ -84,6 +84,31 @@ function run() {
 					'ob_callback'
 				) );
 		}
+
```
Comment (maxicus) [inline Generic_Plugin.php:87]: there's still code in general.

The idea is that if you pollute general module - everything will be there at some point and it becomes unmaintantable.

File: Extension_CloudFlare_Page_View.js (hunk 5 above)
Comment (cssjoe) [reply Extension_CloudFlare_Page_View.js]: I'm fixing all of those "data-" references to see "data()", which works.

File: Generic_Plugin.php (hunk 6 above)
Comment (cssjoe) [reply Generic_Plugin.php]: I moved the code to a new "UserExperience_Plugin" module.

File: Extension_CloudFlare_View_Dashboard.js [hunk 7]
```diff
@@ -1,6 +1,6 @@
 jQuery(function() {
     jQuery('#w3tc_dashboard [type=submit]').bind('click', function(){
-        jQuery(this).attr('was_clicked','yes');
+        jQuery(this).prop('was_clicked','yes');
```
Comment (maxicus) [inline Extension_CloudFlare_View_Dashboard.js:3]: @cssjoe how about that? i guess there are like 100 similar non-default props.

Comment (cssjoe) [reply Extension_CloudFlare_View_Dashboard.js]: I reverted those.

File: Root_Loader.php [hunk 8]
```diff
@@ -20,6 +20,7 @@ function __construct() {
 
 		$plugins = array();
 		$plugins[] = new Generic_Plugin();
+		$plugins[] = new UserExperience_Plugin();
```
Comment (maxicus) [inline Root_Loader.php:23]: this should be conditional to not load files (and waste compilation+execution time and memory) when they are not used.

see e.g. https://github.com/W3EDGE/w3-total-cache/blob/583d7a056f0d19345ac8045ec175c42b4ba414d1/Root_Loader.php#L35

Comment (cssjoe) [reply Root_Loader.php]: I renamed the new class and conditionally loaded it.

**Comment by bwmarkle** [comment]
@cssjoe I'm poking around... The first 2 themes I tested aren't loading jQuery migrate on the front end. Can you simply install any theme and you'll see jquery migrate? How exactly was this tested?

File: UserExperience_Plugin_Jquery.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+	 */
+	public function run() {
+		// Disable jquery-migrate on the front-end, if configured.
+		if ( ! is_admin() && $this->_config->get_boolean( 'jquerymigrate.disabled' ) ) {
+			add_action( 'wp_default_scripts', array( $this, 'disable_jquery_migrate' ) );
+		}
+	}
+
+	/**
+	 * Disable jquery-migrate.
```
Comment (bwmarkle) [inline UserExperience_Plugin_Jquery.php:49]: jQuery migrate is not removed if I enqueue it by itself.

```
add_action(
    'wp_enqueue_scripts',
    function() {
	    wp_enqueue_script( 'jquery-migrate' );
    },
    100
);
```

Current approach only works if it's registered as a dependency.

**Comment by bwmarkle** [comment]
When I add this to the functions.php file of twentyeleven

```php
add_action( 'wp_enqueue_scripts', function() {
    wp_register_script( 'jquery-migrate', '//code.jquery.com/jquery-migrate-1.2.1.min.js', array( 'jquery' ), '1.2.1', true );
    wp_enqueue_script( 'jquery-migrate' );
    wp_enqueue_script( 'jquery' );
} );
```

Then do this:
image

I still see this:
image

**Comment by bwmarkle** [comment]
@cssjoe Any thoughts on these errors?

image

**Comment by cssjoe** [comment]
Check again.  The updated file https://github.com/W3EDGE/w3-total-cache/blob/848ec8faea67e4c2ee73f97dcaaba53630663edc/pub/js/options.js here has `.size()` replaced with `.length`.

**Pull request merged**